WordPress/gutenberg
# Layout Support: Replace incremental IDs with hashes

## What?

Changes how class names for layout containers are generated, replacing per-block, incrementally-generated IDs with hashes obtained from the layout styles definition.

Will fix https://github.com/WordPress/gutenberg/issues/67308.

## Why?

This currently affects the style of elements outside of router regions, even when navigating between pages generated with the same template, because layout class names depend on the number of blocks rendered using them (see https://github.com/WordPress/gutenberg/pull/67826), which can vary.

## How?

The PR adds a new function named `gutenberg_unique_id_from_values`, which basically generates a short hash from the passed array.

This function is used to generate the layout container class, receiving all the data required to generate the style output. That guarantees hashes to be unique for each layout style combination.

Similar examples can be found in the WordPress codebase already:
- [/src/wp-includes/bookmark.php#L154](https://github.com/WordPress/wordpress-develop/blob/f444639e0852f7c0635a09d0dcee561af22a6515/src/wp-includes/bookmark.php#L154)
- [/src/wp-includes/class-wp-oembed-controller.php#L176](https://github.com/WordPress/wordpress-develop/blob/f444639e0852f7c0635a09d0dcee561af22a6515/src/wp-includes/class-wp-oembed-controller.php#L176)
- [/src/wp-includes/block-supports/elements.php#L19](https://github.com/WordPress/wordpress-develop/blob/f444639e0852f7c0635a09d0dcee561af22a6515/src/wp-includes/block-supports/elements.php#L19)
- [/src/wp-includes/global-styles-and-settings.php#L266](https://github.com/WordPress/wordpress-develop/blob/f444639e0852f7c0635a09d0dcee561af22a6515/src/wp-includes/global-styles-and-settings.php#L266)

## Testing Instructions

1. Create a new post.
2. Add four different groups to test each of the different types: flow, constrained, flex and grid. The layout settings can be found in the right-side panel under the "block" tab. Follow these steps to configure each of the types:
   - Flow: disable "Inner blocks use content width".
   - Constrained: enable "Inner blocks use content width".
   - Flex: transform the Group into a Row or a Stack.
   - Grid: transform the Group into a Grid.
3. For the flow and constrained types, you would have to add more settings in order to generate classes with hashes:
   - Flow: add a custom block spacing in the block styles.
   - Constrained: change justification to left or right.
4. Save and visit the post.
5. Ensure the `div` container for these groups has a class name similar to `wp-container-core-group-is-layout-634f0b9d` (a different hash is OK).
6. Edit the post and duplicate all Group elements. Save and visit the post.
7. Ensure the duplicated Group elements have the same hashes as the original Group elements.
8. Edit the post again, modifying some layout settings in the duplicated elements.
9. Ensure the duplicated Group elements now have different hashes.


## Screenshots or screencast 



|Before|After|
|-|-|
|<video width="480" src="https://github.com/user-attachments/assets/da0ba463-7793-46b2-9d5b-500b2f61739a"></video>|<video width="480" src="https://github.com/user-attachments/assets/d5e6a1db-6ccf-46fb-93d7-e2129c7be0d7"></video>|
